File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cpp
#include <vector>
#include <queue>
using namespace std;

int mv1[4] = {-1, 1, 0, 0};
int mv2[4] = {0, 0, -1, 1};

int solution(vector<vector<int> > maps){
    queue<pair<int, int>> q;
    q.push({0, 0});
    
    while(!q.empty()){
        int x = q.front().first;
        int y = q.front().second;
        q.pop();
        
        for(int i = 0; i < 4; i++){
            // 상 하 좌 우 순서로 진행
            int nx = x + mv1[i]; 
            int ny = y + mv2[i];
            
            if((nx >= 0 && nx < maps.size()) && (ny >= 0 && ny < maps[0].size())){ // 범위가 넘어가면 무시
                if(maps[nx][ny] == 1){
                    maps[nx][ny] = maps[x][y] + 1;
                    q.push({nx, ny});
                }
            }
            
            if(nx == maps.size() - 1 && ny == maps[0].size() - 1) return maps[nx][ny];
        }
    }
    
    return -1;
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Pair {
    int x, y, cnt;
    Pair(int x, int y, int cnt) {
        this.x = x;
        this.y = y;
        this.cnt = cnt;
    }
}

class Solution {
    int mv[][] = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
    
    public int BFS(int[][] maps) {
        Queue<Pair> q = new LinkedList<>();
        q.add(new Pair(0, 0, 1));
        maps[0][0] = 0;
        
        int ex = maps.length, ey = maps[0].length;
        
        while(!q.isEmpty()) {
            int x = q.peek().x;
            int y = q.peek().y;
            int cnt = q.peek().cnt;
            q.remove();
            
            if(x == ex - 1 && y == ey - 1) return cnt;
            
            for(int i = 0; i < 4; i++) {
                int nx = x + mv[i][0];
                int ny = y + mv[i][1];
                
                if(nx < 0 || ny < 0 || nx >= ex || ny >= ey) continue;
                if(maps[nx][ny] == 0) continue;
                
                maps[nx][ny] = 0;
                q.add(new Pair(nx, ny, cnt + 1));
            }
        }
        
        return -1;
    }
    
    public int solution(int[][] maps) {
        return BFS(maps);
    }
}